internal/resolver/delegatingresolver/delegatingresolver.go

```diff
@@ -0,0 +1,241 @@
+// resolvers and the gRPC ClientConn. This type implements the
+// resolver.ClientConn interface to intercept state updates from both the target
+// and proxy resolvers, combining the results before forwarding them to the
+// client connection.
+type delegatingResolver struct {
+	target         resolver.Target     // parsed target URI to be resolved
+	cc             resolver.ClientConn // gRPC ClientConn
+	targetResolver resolver.Resolver   // resolver for the target URI, based on its scheme
+	proxyResolver  resolver.Resolver   // resolver for the proxy URI; nil if no proxy is configured
+	targetAddrs    []resolver.Address  // resolved or unresolved target addresses, depending on proxy configuration
+	proxyAddrs     []resolver.Address  // resolved proxy addresses; empty if no proxy is configured
+	proxyURL       *url.URL            // proxy URL, derived from proxy environment and target
+	mu             sync.Mutex          // protects access to the resolver state during updates
+}
+
+func mapAddress(address string) (*url.URL, error) {
+	req := &http.Request{
+		URL: &url.URL{
+			Scheme: "https",
+			Host:   address,
+		},
+	}
+	url, err := httpProxyFromEnvironment(req)
+	if err != nil {
+		return nil, err
+	}
+	return url, nil
+}
+
+// New creates a new delegating resolver that will call the target and proxy
+// child resolvers based on the proxy enviorinment configurations.
+func New(target resolver.Target, cc resolver.ClientConn, opts resolver.BuildOptions, targetResolverBuilder resolver.Builder, targetResolutionEnabled bool) (resolver.Resolver, error) {
+	r := &delegatingResolver{
+		target: target,
+		cc:     cc,
+	}
+
+	var err error
+	r.proxyURL, err = mapAddress(target.Endpoint())
+	if err != nil {
+		return nil, fmt.Errorf("failed to determine proxy URL for target : %v", err)
+	}
+
+	if r.proxyURL == nil {
+		logger.Info("No proxy URL detected")
+		return targetResolverBuilder.Build(target, cc, opts)
+	}
+
+	// When the scheme is 'dns', resolution should be handled by the proxy, not
+	// the client. Therefore, we bypass the target resolver and store the
+	// unresolved target address.
+	if target.URL.Scheme == "dns" && !targetResolutionEnabled {
+		r.targetAddrs = []resolver.Address{{Addr: target.Endpoint()}}
+	} else {
+		// targetResolverBuilder must not be nil. If it is nil, the channel should
+		// error out and not call this function.
+		r.targetResolver, err = targetResolverBuilder.Build(target, &wrappingClientConn{r, "target"}, opts)
+		if err != nil {
+			return nil, fmt.Errorf("delegating_resolver: unable to build the target resolver : %v", err)
+		}
+	}
+	r.proxyResolver, err = proxyURIResolver(r.proxyURL, opts, r)
+	if err != nil {
+		return nil, err
+	}
+	return r, nil
+}
+
+func proxyURIResolver(proxyURL *url.URL, opts resolver.BuildOptions, presolver *delegatingResolver) (resolver.Resolver, error) {
+	scheme := "dns"
+	proxyBuilder := resolver.Get(scheme)
```

Comment:
should it be presolver? Also, this function can be part of struct implementation i.e. `func (r *delegatingResolver)` so that you can access the resolvers and opts without passing here.

Reply:
Done